File: examples/catadioptre-java-gradle-groovy-dsl-example/src/test/java/io/aerisconsulting/catadioptre/example/CatadioptreExampleTest.java
package io.aerisconsulting.catadioptre.example;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

class CatadioptreExampleTest {

	@Test
	@DisplayName("should read the value of default property")
	void shouldReadTheValueOfDefaultProperty() {
		Map<String, Double> defaultProperty = new HashMap<>();
		defaultProperty.put("any", 1.0);
		CatadioptreExample instance = new CatadioptreExample(defaultProperty, 1.0, Optional.empty());

		Map<String, Double> result = TestableCatadioptreExample.markers(instance);

		Assertions.assertThat(result).isNotNull().hasSize(1).containsEntry("any", 1.0);
	}

	@Test
	@DisplayName("should write the value of default property")
	void shouldWriteTheValueOfDefaultProperty() {
		Map<String, Double> defaultProperty = new HashMap<>();
		defaultProperty.put("any", 1.0);
		CatadioptreExample instance = new CatadioptreExample(defaultProperty, 1.0, Optional.empty());

		defaultProperty = new HashMap<>();
		defaultProperty.put("other", 2.0);
		CatadioptreExample result = TestableCatadioptreExample.markers(instance, defaultProperty);

		Assertions.assertThat(result).isSameAs(instance);
		Assertions.assertThat(TestableCatadioptreExample.markers(result)).isNotNull().hasSize(1)
				.containsEntry("other", 2.0);
	}

	@Test
	@DisplayName("should clear the value of default property")
	void shouldClearTheValueOfDefaultProperty() {
		Map<String, Double> defaultProperty = new HashMap<>();
		defaultProperty.put("any", 1.0);
		CatadioptreExample instance = new CatadioptreExample(defaultProperty, 1.0, Optional.empty());

		CatadioptreExample result = TestableCatadioptreExample.clearMarkers(instance);

		Assertions.assertThat(result).isSameAs(instance);
		Assertions.assertThat(TestableCatadioptreExample.markers(result)).isNull();
	}

	@Test
	@DisplayName("should write and read the value of inherited typed parameter as double")
	void shouldWriteAndReadTheValueOfInheritedTypedParameterAsDouble() {
		Map<String, Double> defaultProperty = new HashMap<>();
		defaultProperty.put("any", 1.0);
		CatadioptreExample instance = new CatadioptreExample(defaultProperty, 1.0, Optional.empty());

		CatadioptreExample result = TestableAbstractCatadioptreExample.typedProperty(instance, 2.0);

		Assertions.assertThat(result).isSameAs(instance);
		Assertions.assertThat(TestableAbstractCatadioptreExample.typedProperty(result)).isEqualTo(2.0);

	}

	@Test
	@DisplayName("should write and read the value of inherited typed parameter as string")
	void shouldWriteAndReadTheValueOfInheritedTypedParameterAsString() {
		Map<String, Double> defaultProperty = new HashMap<>();
		defaultProperty.put("any", 1.0);
		CatadioptreExample instance = new CatadioptreExample(defaultProperty, 1.0, Optional.empty());

		CatadioptreExample result = TestableAbstractCatadioptreExample.typedProperty2(instance,
				Optional.of("this is the value"));

		Assertions.assertThat(result).isSameAs(instance);
		Assertions.assertThat(TestableAbstractCatadioptreExample.typedProperty2(result))
				.isEqualTo(Optional.of("this is the value"));
	}

	@Test
	@DisplayName("should execute the function without argument")
	void shouldExecuteTheFunctionWithoutArgument() {
		Map<String, Double> defaultProperty = new HashMap<>();
		defaultProperty.put("any", 1.0);
		CatadioptreExample instance = new CatadioptreExample(defaultProperty, 1.0, Optional.empty());

		double result = TestableCatadioptreExample.multiplySum(instance, 2.0, new Double[]{1.0, 3.0, 6.0});

		Assertions.assertThat(result).isEqualTo(20.0);
	}

	@Test
	@DisplayName("should execute the function of parent without argument")
	void shouldExecuteTheFunctionOfParentWithoutArgument() {
		Map<String, Double> defaultProperty = new HashMap<>();
		defaultProperty.put("any", 1.0);
		CatadioptreExample instance = new CatadioptreExample(defaultProperty, 1.0, Optional.empty());

		int result = TestableAbstractCatadioptreExample.getAnything(instance);

		Assertions.assertThat(result).isEqualTo(123);
	}

	@Test
	@DisplayName("should execute the function of parent with variable arguments")
	void shouldExecuteTheFunctionOfParentWithVariableArguments() {
		Map<String, Double> defaultProperty = new HashMap<>();
		defaultProperty.put("any", 1.0);
		CatadioptreExample instance = new CatadioptreExample(defaultProperty, 1.0, Optional.empty());

		double result = TestableAbstractCatadioptreExample.divideSum(instance, 2.0, new Double[]{1.0, 3.0, 6.0});

		Assertions.assertThat(result).isEqualTo(5.0);
	}
}

File: examples/catadioptre-kotlin-gradle-groovy-dsl-example/src/test/kotlin/io/aerisconsulting/catadioptre/example/CatadioptreExampleTest.kt
/*
 * Copyright 2021 AERIS-Consulting e.U.
 *
 * AERIS-Consulting e.U. licenses this file to you under the Apache License, version 2.0 (the
 * "License"); you may not use this file except in compliance with the License. You may obtain a
 * copy of the License at:
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package io.aerisconsulting.catadioptre.example

import assertk.assertThat
import assertk.assertions.isEqualTo
import assertk.assertions.isNull
import assertk.assertions.isSameAs
import io.aerisconsulting.catadioptre.example.catadioptre.clearDefaultProperty
import io.aerisconsulting.catadioptre.example.catadioptre.defaultArgumentTypedProperty
import io.aerisconsulting.catadioptre.example.catadioptre.defaultArgumentTypedProperty2
import io.aerisconsulting.catadioptre.example.catadioptre.defaultProperty
import io.aerisconsulting.catadioptre.example.catadioptre.divideSum
import io.aerisconsulting.catadioptre.example.catadioptre.getAnything
import io.aerisconsulting.catadioptre.example.catadioptre.multiplySum
import org.junit.jupiter.api.Test
import java.util.Optional

internal class CatadioptreExampleTest {

    @Test
    internal fun `should read the value of default property`() {
        val instance = CatadioptreExample()

        val result = instance.defaultProperty()

        assertThat(result).isEqualTo(mapOf("any" to 1.0))
    }

    @Test
    internal fun `should write the value of default property`() {
        val instance = CatadioptreExample()

        val result = instance.defaultProperty(mapOf("other" to 2.0))

        assertThat(result).isSameAs(instance)
        assertThat(instance.defaultProperty()).isEqualTo(mapOf("other" to 2.0))
    }

    @Test
    internal fun `should clear the value of nullable default property`() {
        val instance = CatadioptreExample()

        val result = instance.clearDefaultProperty()

        assertThat(result).isSameAs(instance)
        assertThat(instance.defaultProperty()).isNull()
    }

    @Test
    internal fun `should write and read the value of inherited typed parameter as double`() {
        val instance = CatadioptreExample()

        val result = instance.defaultArgumentTypedProperty(2.0)

        assertThat(result).isSameAs(instance)
        assertThat(instance.defaultArgumentTypedProperty()).isEqualTo(2.0)
    }

    @Test
    internal fun `should write and read the value of inherited typed parameter as string`() {
        val instance = CatadioptreExample()

        val result = instance.defaultArgumentTypedProperty2(Optional.of("this is the value"))

        assertThat(result).isSameAs(instance)
        assertThat(instance.defaultArgumentTypedProperty2()).isEqualTo(Optional.of("this is the value"))
    }

    @Test
    internal fun `should execute the function without argument`() {
        val instance = CatadioptreExample()

        val result = instance.multiplySum(2.0, 1.0, 3.0, 6.0)

        assertThat(result).isEqualTo(20.0)
    }

    @Test
    internal fun `should execute the function of parent without argument`() {
        val instance = CatadioptreExample()

        val result = instance.getAnything()

        assertThat(result).isEqualTo(123)
    }

    @Test
    internal fun `should execute the function of parent with variable arguments`() {
        val instance = CatadioptreExample()

        val result = instance.divideSum(2.0, 1.0, 3.0, 6.0)

        assertThat(result).isEqualTo(5.0)
    }

}
